src/renderers/shared/fiber/ReactFiber.js

```diff
@@ -86,6 +117,33 @@ function shouldConstruct(Component) {
   return !!(Component.prototype && Component.prototype.isReactComponent);
 }
 
+// This is used to create an alternate fiber to do work on.
+exports.cloneFiber = function(fiber : Fiber) : Fiber {
+  // We use a double buffering pooling technique because we know that we'll only
+  // ever need at most two versions of a tree. We pool the "other" unused node
```

Comment:
If we build a test where this happens, I think we're likely to fail the test anyway.

It is true that this is super difficult to reason about and I think there might even be a problem in the code right now because of the use of "alternate" in ReactChildFiber. To make this work without the pooling, I'd have to fix those callsites. I've been looking at it last night but it is harder than it seems because the return value for "next" becomes a tuple if you want to code completely without the use of it.

That means that alternate is also acting as "current" version.

It would be nice to assert dev only but not sure how to do that efficiently. Do you have any ideas?